pkg/controller/deployment/sync.go

```diff
@@ -46,6 +47,104 @@ func (dc *DeploymentController) syncStatusOnly(deployment *extensions.Deployment
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// used for unit testing
+var nowFn = func() time.Time { return time.Now() }
+
+// hasTimedOut let us know if a deployment did not manage to finish before the specified
+// timeout (progressDeadlineSeconds). The controller should stop processing it and in a
+// follow-up we should enable automatically rollback to the previous successful replica
+// set.
+func (dc *DeploymentController) hasTimedOut(deployment *extensions.Deployment) (bool, error) {
+	if deployment.Spec.ProgressDeadlineSeconds == nil {
+		return false, nil
+	}
+
+	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(deployment, false)
+	if err != nil {
+		return false, err
+	}
+	allRSs := oldRSs
+
+	// On template changes we should try to cleanup the deployment status and not look at
+	// the timeout.
+	if newRS == nil {
+		// Remove any condition that denotes lack of progress for a new rollout is underway.
+		removed := deploymentutil.RemoveCondition(deployment, extensions.DeploymentProgressing, api.ConditionFalse)
+		if !removed {
+			return false, nil
+		}
+
+		d, err := dc.client.Extensions().Deployments(deployment.Namespace).UpdateStatus(deployment)
```

Comment:
Yes, see the `worker` in deployment_controller.go. It calls handleErr for every error returned from `syncDeployment`.